# Dimentionality Reduction

### Load the cleaned transaction-level data.

In [1]:
# NumPy for numerical computing
import numpy as np

# Pandas for DataFrames
import pandas as pd
pd.set_option('display.max_columns', 100)

# Matplotlib for visualization
from matplotlib import pyplot as plt
# display plots in the notebook
%matplotlib inline

# Seaborn for easier visualization
import seaborn as sns

In [2]:
df = pd.read_csv('cleaned_transactions.csv')

### Item data

In [3]:
# First 10 StockCodes and Descriptions
df[['StockCode', 'Description']].head(10)

StockCode                          Description
0     22728            ALARM CLOCK BAKELIKE PINK
1     22727            ALARM CLOCK BAKELIKE RED 
2     22726           ALARM CLOCK BAKELIKE GREEN
3     21724      PANDA AND BUNNIES STICKER SHEET
4     21883                     STARS GIFT TAPE 
5     10002          INFLATABLE POLITICAL GLOBE 
6     21791   VINTAGE HEADS AND TAILS CARD GAME 
7     21035      SET/2 RED RETROSPOT TEA TOWELS 
8     22326  ROUND SNACK BOXES SET OF4 WOODLAND 
9     22629                  SPACEBOY LUNCH BOX

In [4]:
# number of unique items
len(df.StockCode.unique())

2574

Problem: we will have 2574 dummy variables, thus there will be very high dimendionality.

### High dimensionality

(create a dataframe of dummy variables for 'StockCode'; add CustomerID)

In [5]:
df.head()

InvoiceNo StockCode                      Description  Quantity  \
0     536370     22728        ALARM CLOCK BAKELIKE PINK        24   
1     536370     22727        ALARM CLOCK BAKELIKE RED         24   
2     536370     22726       ALARM CLOCK BAKELIKE GREEN        12   
3     536370     21724  PANDA AND BUNNIES STICKER SHEET        12   
4     536370     21883                 STARS GIFT TAPE         24   

    InvoiceDate  UnitPrice  CustomerID Country  Sales  
0  12/1/10 8:45       3.75       12583  France   90.0  
1  12/1/10 8:45       3.75       12583  France   90.0  
2  12/1/10 8:45       3.75       12583  France   45.0  
3  12/1/10 8:45       0.85       12583  France   10.2  
4  12/1/10 8:45       0.65       12583  France   15.6

In [6]:
# item_dummies
item_dummies = pd.get_dummies(df.StockCode)

# add CustomerID to item_dummies
item_dummies['CustomerID'] = df.CustomerID

# first 5 rows of item_dummies
item_dummies.head()

10002  10120  10125  10133  10135  11001  15034  15036  15039  15044A  \
0      0      0      0      0      0      0      0      0      0       0   
1      0      0      0      0      0      0      0      0      0       0   
2      0      0      0      0      0      0      0      0      0       0   
3      0      0      0      0      0      0      0      0      0       0   
4      0      0      0      0      0      0      0      0      0       0   

   15044B  15044C  15044D  15056BL  15056N  15056P  15058A  15058B  15058C  \
0       0       0       0        0       0       0       0       0       0   
1       0       0       0        0       0       0       0       0       0   
2       0       0       0        0       0       0       0       0       0   
3       0       0       0        0       0       0       0       0       0   
4       0       0       0        0       0       0       0       0       0   

   15060B  16008  16011  16012  16014  16016  16045  16048  16054  16156L  \
0       0      0      0      0      0      0      0      0      0       0   
1       0      0      0      0      0      0      0      0      0       0   
2       0      0      0      0      0      0      0      0      0       0   
3       0      0      0      0      0      0      0      0      0       0   
4       0      0      0      0      0      0      0      0      0       0   

   16156S  16161G  16161P  16161U  16168M  16169E  16169K  16169M  16218  \
0       0       0       0       0       0       0       0       0      0   
1       0       0       0       0       0       0       0       0      0   
2       0       0       0       0       0       0       0       0      0   
3       0       0       0       0       0       0       0       0      0   
4       0       0       0       0       0       0       0       0      0   

   16219  16225  16235  16236  16237  16238  16258A  16259  17003  17011F  \
0      0      0      0      0      0      0       0      0      0       0   
1      0      0      0      0      0      0       0      0      0       0   
2      0      0      0      0      0      0       0      0      0       0   
3      0      0      0      0      0      0       0      0      0       0   
4      0      0      0      0      0      0       0      0      0       0   

   17012A  17012B     ...      85232A  85232B  85232D  90001B  90001D  90013A  \
0       0       0     ...           0       0       0       0       0       0   
1       0       0     ...           0       0       0       0       0       0   
2       0       0     ...           0       0       0       0       0       0   
3       0       0     ...           0       0       0       0       0       0   
4       0       0     ...           0       0       0       0       0       0   

   90013C  90018C  90019A  90024B  90030A  90030B  90030C  90031  90036E  \
0       0       0       0       0       0       0       0      0       0   
1       0       0       0       0       0       0       0      0       0   
2       0       0       0       0       0       0       0      0       0   
3       0       0       0       0       0       0       0      0       0   
4       0       0       0       0       0       0       0      0       0   

   90037B  90057  90070  90087  90098  90099  90108  90114  90120B  90145  \
0       0      0      0      0      0      0      0      0       0      0   
1       0      0      0      0      0      0      0      0       0      0   
2       0      0      0      0      0      0      0      0       0      0   
3       0      0      0      0      0      0      0      0       0      0   
4       0      0      0      0      0      0      0      0       0      0   

   90160A  90160B  90160C  90160D  90161B  90161C  90161D  90162A  90162B  \
0       0       0       0       0       0       0       0       0       0   
1       0       0       0       0       0       0       0       0       0   
2       0       0       0       0       0       0       

### Roll up the item dummies data into customer-level item data.

In [7]:
# aggregating at customer level
item_data = item_dummies.groupby('CustomerID').sum()

# first 5 rows of item_data
item_data.head()

10002  10120  10125  10133  10135  11001  15034  15036  15039  \
CustomerID                                                                  
12347           0      0      0      0      0      0      0      0      0   
12348           0      0      0      0      0      0      0      0      0   
12349           0      0      0      0      0      0      0      0      0   
12350           0      0      0      0      0      0      0      0      0   
12352           0      0      0      0      0      0      0      0      0   

            15044A  15044B  15044C  15044D  15056BL  15056N  15056P  15058A  \
CustomerID                                                                    
12347            0       0       0       0        0       0       0       0   
12348            0       0       0       0        0       0       0       0   
12349            0       0       0       0        0       0       0       0   
12350            0       0       0       0        0       0       0       0   
12352            0       0       0       0        0       0       0       0   

            15058B  15058C  15060B  16008  16011  16012  16014  16016  16045  \
CustomerID                                                                     
12347            0       0       0      1      0      0      0      0      0   
12348            0       0       0      0      0      0      0      0      0   
12349            0       0       0      0      0      0      0      0      0   
12350            0       0       0      0      0      0      0      0      0   
12352            0       0       0      0      0      0      0      0      0   

            16048  16054  16156L  16156S  16161G  16161P  16161U  16168M  \
CustomerID                                                                 
12347           0      0       0       0       0       0       0       0   
12348           0      0       0       0       0       0       0       0   
12349           0      0       0       0       0       0       0       0   
12350           0      0       0       0       0       0       0       0   
12352           0      0       0       0       0       0       0       0   

            16169E  16169K  16169M  16218  16219  16225  16235  16236  16237  \
CustomerID                                                                     
12347            0       0       0      0      0      0      0      0      0   
12348            0       0       0      0      0      0      0      0      0   
12349            0       0       0      0      0      0      0      0      0   
12350            0       0       0      0      0      0      0      0      0   
12352            0       0       0      0      0      0      0      0      0   

            16238  16258A  16259  17003  17011F  17012A  17012B  ...   85231B  \
CustomerID                                                       ...            
12347           0       0      0      0       0       0       0  ...        0   
12348           0       0      0      0       0       0       0  ...        0   
12349           0       0      0      0       0       0       0  ...        0   
12350           0       0      0      0       0       0       0  ...        0   
12352           0       0      0      0       0       0       0  ...        0   

            85232A  85232B  85232D  90001B  90001D  90013A  90013C  90018C  \
CustomerID                                                                   
12347            0       0       1       0       0       0       0       0   
12348            0       0       0       0       0       0       0       0   
12349            0       0       0       0       0       0       0       0   
12350            0       0       0       0       0       0       0       0   
12352            0       0       0       0       0       0       0       0   

            90019A  90024B  90030A  90030B  90030C  90031  90036E  90037B  \
CustomerID                                                             

In [8]:
# total times each item was purchased
item_data.sum()

10002        12
10120         1
10125        13
10133         5
10135         4
11001         8
15034         5
15036        19
15039         3
15044A        6
15044B        3
15044C        2
15044D        4
15056BL      50
15056N       35
15056P       24
15058A        9
15058B        8
15058C        4
15060B       12
16008        11
16011         3
16012         4
16014        10
16016        16
16045         8
16048         8
16054         2
16156L        6
16156S       12
           ... 
90098         1
90099         2
90108         1
90114         1
90120B        1
90145         2
90160A        1
90160B        1
90160C        1
90160D        1
90161B        1
90161C        1
90161D        1
90162A        1
90162B        1
90164A        1
90170         1
90173         1
90184B        1
90184C        1
90192         1
90201A        1
90201B        3
90201C        2
90201D        1
90202D        1
90204         1
C2            6
M            34
POST       1055
Length: 2574, dtype: int

### Save item_data

In [9]:
item_data.to_csv('item_data.csv')

### Thresholds

(getting only the top items to reduce dimensionality)

In [10]:
# most popular 20 items
item_data.sum().sort_values().tail(20)

22961      114
22630      115
22139      117
21080      122
85099B     123
20726      123
20719      128
20750      132
23084      140
20725      141
21212      143
22551      158
22629      160
22328      166
21731      169
22556      179
22554      197
22423      222
22326      271
POST      1055
dtype: int64

In [11]:
# get list of StockCodes for the 20 most popular items
top_20_items = item_data.sum().sort_values().tail(20).index

top_20_items

Index(['22961', '22630', '22139', '21080', '85099B', '20726', '20719', '20750',
       '23084', '20725', '21212', '22551', '22629', '22328', '21731', '22556',
       '22554', '22423', '22326', 'POST'],
      dtype='object')

Keep only the features for those 20 items.

In [12]:
top_20_item_data = item_data[top_20_items]

# shape
top_20_item_data.shape

(414, 20)

In [13]:
top_20_item_data.head()

22961  22630  22139  21080  85099B  20726  20719  20750  23084  \
CustomerID                                                                   
12347           0      0      0      0       0      0      4      0      3   
12348           0      0      0      0       0      0      0      0      0   
12349           0      0      0      0       0      0      0      0      0   
12350           0      0      0      0       0      0      0      0      0   
12352           0      1      0      0       0      0      0      0      0   

            20725  21212  22551  22629  22328  21731  22556  22554  22423  \
CustomerID                                                                  
12347           0      0      0      0      0      5      0      0      4   
12348           0      0      0      0      0      0      0      0      0   
12349           0      0      0      0      0      0      1      1      1   
12350           0      0      1      0      0      0      0      0      0   
12352           0      0      0      0      0      1      0      0      2   

            22326  POST  
CustomerID               
12347           0     0  
12348           0     4  
12349           1     1  
12350           0     1  
12352           0     5

### Save the threshold item data

In [14]:
# Save threshold_item_data.csv
top_20_item_data.to_csv('threshold_item_data.csv')